File: 74751_lab4_task5.cpp
#include <iostream>
#include <vector>
#include <string>

using namespace std;

vector<double> apply_filter(const vector<double>* original_signal, const vector<double>* filter_coefficients) {
    vector<double> filtered;
    int n = original_signal->size();
    int m = filter_coefficients->size();
    
    for (int i = 0; i < n; ++i) {
        double val = 0;
        for (int j = 0; j < m; ++j) {
            if (i - j >= 0) {
                val += (*original_signal)[i - j] * (*filter_coefficients)[j];
            }
        }
        filtered.push_back(val);
    }
    return filtered;
}

int process_log(string* log_str, string target_symptom) {
    int count = 0;
    size_t pos = 0;
    while ((pos = log_str->find(target_symptom, pos)) != string::npos) {
        count++;
        string upper_target = target_symptom;
        for (char& c : upper_target) {
            c = toupper(c);
        }
        log_str->replace(pos, target_symptom.length(), upper_target);
        pos += upper_target.length();
    }
    return count;
}

int main() {
    vector<double> sig = {1, 2, 3, 4, 5};
    vector<double> coeffs = {0.2, 0.5, 0.2};
    vector<double> filtered_sig = apply_filter(&sig, &coeffs);
    cout << "Filtered signal: ";
    for (double v : filtered_sig) cout << v << " ";
    cout << endl;
    
    string log = "headache, nausea, headache, fever";
    int count = process_log(&log, "headache");
    cout << "Count: " << count << ", Modified log: " << log << endl;
    
    return 0;
}

File: 74751_lab4_task5.py
def apply_filter(original_signal, filter_coefficients):
    filtered = []
    for i in range(len(original_signal)):
        val = 0
        for j in range(len(filter_coefficients)):
            if i - j >= 0:
                val += original_signal[i - j] * filter_coefficients[j]
        filtered.append(val)
    return filtered

def process_log(log_ref, target_symptom):
    # Using a list containing a single string to simulate pointer/reference behavior
    log_str = log_ref[0]
    count = log_str.count(target_symptom)
    # Modify log string by replacing the symptom with uppercase
    log_ref[0] = log_str.replace(target_symptom, target_symptom.upper())
    return count

if __name__ == "__main__":
    sig = [1, 2, 3, 4, 5]
    coeffs = [0.2, 0.5, 0.2]
    filtered_sig = apply_filter(sig, coeffs)
    print(f"Filtered signal: {filtered_sig}")
    
    log = ["headache, nausea, headache, fever"]
    count = process_log(log, "headache")
    print(f"Count: {count}, Modified log: {log[0]}")
